feat(uraft): Introduce floating IP manager for HA

ralcolea opened the pull request:
Previously, a floating IP drop triggered a new election in uRaft, sometimes leading to leader flip-flopping.

This PR introduces a FloatingIP Manager to efficiently handle floating IPs using an AF_NETLINK-based event listener. This ensures low CPU usage, automatic IP failover, and continuous monitoring.

**Key Features & Changes**
- **FloatingIP Manager Implementation**
   - Manages the floating IP lifecycle.
   - Provides automatic IP restoration on failure.
   - Supports clean start/stop functionality.

- **Netlink Event Listener**
  - AF_NETLINK + select() approach for real-time event detection.
  - Monitors RTM_DELADDR events to detect IP removal.
  - Captures network interface & IP address for accurate recovery.
  - Efficient, non-blocking design with adjustable timeout (default value 500 ms).

- **High Availability Enhancements**
  - Auto-recovery mechanism: If the socket crashes, it is restarted automatically.
  - Stop flag mechanism to gracefully exit the listener thread.
  - Failover handling: If the floating IP is lost, saunafs-uraft-helper script is used to restore the floating IP.

- **Code Quality & Performance**
  - Optimized low CPU consumption compared to polling-based methods.
  - Improved error handling and logging.
  - Added unit tests for uRaft (floating_ip_manager_unittest.cc) and integration tests in Ansible repository (test-uraft-floating-ip-manager branch).

**Advantages of this change**
- Replaces inefficient polling (arping/ip addr show) with a real-time, event-driven approach.
- Ensures high availability by detecting IP loss instantly.
 
🚀 **Performance Improvements**
✅ Netlink approach: Low CPU overhead, instant wakes up during IP drop.
❌ Polling (arping or ip addr show): High CPU usage, unnecessary checks.

🛠️ Testing Done
✅ Unit-tested IP detection & restoration.
✅ Simulated IP removal events.

ralcolea committed:
feat(uraft): Introduce floating IP manager for HA

Previously, a floating IP drop triggered a new election in uRaft,
sometimes leading to leader flip-flopping.

This commit introduces a Floating IP manager to improve high
availability by actively monitoring and managing floating IP
availability. It uses a socket-based event listener to detect
floating IP loss and initiates recovery actions to restore it.

Co-authored-by: Piotr Modrzyk <piotr.modrzyk@gmail.com>

Signed-off-by: Crash <crash@leil.io>

lgsilva3087 commented on src/uraft/floating-ip-manager.h:
```diff
@@ -0,0 +1,134 @@
+#pragma once
+
+#ifndef FLOATING_IP_MANAGER_H
+#define FLOATING_IP_MANAGER_H
```
Please use only one option to ensure the header is included once.
The project usually uses:
`#pragma once`

ralcolea replied:
I will use `#pragma once`.

lgsilva3087 commented on src/uraft/floating-ip-manager.h:
```diff
@@ -0,0 +1,134 @@
+	int checkFloatingIPPeriodMS = 500; ///< Period (in milliseconds) to check the floating IP status.
+};
+
+#endif  // FLOATING_IP_MANAGER_H
```
Maybe not needed according to my previous comment.

ralcolea replied:
Done.

lgsilva3087 commented on src/uraft/floating-ip-manager.h:
```diff
@@ -0,0 +1,134 @@
+
+protected:
+	std::thread listenerThread; ///< Background thread for event listening.
+	std::atomic<bool> stopListenerFlag{false}; ///< Flag to stop the event listener thread.
```
Consider std::jthread that is designed for this pattern.

ralcolea replied:
Very good suggestion. Done.

lgsilva3087 commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+				if (nlh->nlmsg_type == NLMSG_DONE) { break; }
+				if (nlh->nlmsg_type == NLMSG_ERROR) {
+					syslog(LOG_WARNING,
+					       "[FloatingIPManager] Received an error message.");
```
Does the message contains printable text?

ralcolea replied:
Yes, it does. I added the error information for this scenario.

lgsilva3087 commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+			continue;  // Retry loop
+		}
+
+		while (!stopListenerFlag) {
```
Consider to move this loop to a new function, to separate it from the retry logic.
The new function name could be pollForIpRemovalEvents.

ralcolea replied:
Very nice suggestion. After applying this suggestion, function reponsibilities are better distributed and the code is more readable and modular.

lgsilva3087 reviewed (approved):
Great work. Please see my suggestions.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	return (result == 0);
+}
```
In the future, I think we should move more of the bash script stuff into the C++ code. The bash script can be notoriously slow (speed is a requirement for a program like uraft) and we should generally not rely on an outside script (however tempting it may be). Anything we can do in the bash script, we can do in C++.

However, I like that you are keeping things simple right now, and using the script. Gradually we could phase-out the script, but this is perfect start.

However, please be careful when using standard C-library functions, if these functions fail, they will often set errno which will contain actual information on the error. Worse, if you call any other function between this time and, it might change errno. See `man 3 errno` and `man system`

Please change this to actually print what error occurred using errno and strerror. read `man 3 strerror` for a thread safe version.

ralcolea replied:
I agree with you and I really like the idea of moving most of the bash script functions into the C++ code. However, as it will take more time and the priority is to solve the leader flip-flopping issue, I decided to keep the bash script in its current state and then start moving progressively its functions to the C++ code.

I have been addressing your suggestions related with the use of C-library functions and I improved the errors logging based on `errno` as you suggested. Please, take a look again and let me know if the current version is more aligned to your expectations.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+			       kRetryTimeout);
+			std::this_thread::sleep_for(std::chrono::seconds(kRetryTimeout));
+			continue;  // Retry loop
+		}
```
As above, read `man 3 socket`. Usually standard C library functions set errno if they return less than 0, however you need to check and confirm it using `man 3 <FUNCTION_NAME>`

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+			close(sock_fd);
+			std::this_thread::sleep_for(std::chrono::seconds(kRetryTimeout));
+			continue;  // Retry loop
+		}
```
Check if errno is set and print out actual error (see above)

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+				syslog(LOG_WARNING,
+				       "[FloatingIPManager] select() function failed.");
+				break;
+			}
```
Check if errno is set and print out actual error (see above)

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+				       "[FloatingIPManager] recvmsg() failed. "
+				       "Restarting socket...");
+				break;  // Restart socket
+			}
```
Check if errno is set and print out actual error (see above)

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+			}
+		}
+
+		close(sock_fd);
```
Check for error

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+						       "interface name for index: %d",
+						       ifa->ifa_index);
+						continue;  // It happens when the IP is already removed
+					}
```
errno is set on error, print out the actual error (see above)

> // It happens when the IP is already removed

Are you sure about this? This doesn't seem to have anything to do with IP addresses, only interfaces.
My concern here is that interfaces could be removed. This might be the only case were a leader demote may be warranted (as there is no interface, no way to actually bind to anything).

ralcolea replied:
The documentation mentions as possible errors `ENODEV: No interface found with give name` and `ENXIO: No interface found for the index`. But, it also says that the function can fail for any of the errors specified by `socket()` or `ioctl()`. I agree with you. This error does not seem related with IP addresses, but with interfaces.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+						if (rta->rta_type == IFA_LOCAL) { // Local IP address
+							char ipAddress[INET_ADDRSTRLEN] = {0};
+							inet_ntop(AF_INET, RTA_DATA(rta), ipAddress,
+							          sizeof(ipAddress));
```
According to `man 3 inet_ntop`, this function will return NULL and set errno in case of error.

ralcolea replied:
Done.

uristdwarf commented on src/uraft/floating_ip_manager_unittest.cc:
```diff
@@ -0,0 +1,165 @@
+protected:
+	void SetUp() override {
+		// Create a virtual network interface for testing
+		if (system("sudo ip link add name test0 type dummy") != 0) {
+			std::cerr << "Failed to create test0 network interface\n";
+			std::exit(EXIT_FAILURE);
+		}
+		if (system("sudo ip addr add 192.168.1.100/24 dev test0") != 0) {
+			std::cerr << "Failed to add 192.168.1.100 to test0\n";
+			std::exit(EXIT_FAILURE);
+		}
+		if (system("sudo ip addr add 192.168.1.105/24 dev test0") != 0) {
+			std::cerr << "Failed to add 192.168.1.105 to test0\n";
+			std::exit(EXIT_FAILURE);
+		}
+		if (system("sudo ip link set test0 up") != 0) {
+			std::cerr << "Failed to set test0 interface up\n";
+			std::exit(EXIT_FAILURE);
+		}
```
Is sudo needed here? I don't think it will work in the Google Test Framework. Regardless, we are running the tests as root, so you don't need it.

ralcolea replied:
sudo is needed for these tests because, until now, unit tests were not being run as root or with sudo permissions.

From my understanding, using sudo in the Google Test framework works as long as the user running the tests has sudo privileges. IMHO, it's preferable to run only the necessary tests with sudo rather than executing all tests with elevated privileges just because a few require it.

uristdwarf commented on src/uraft/floating_ip_manager_unittest.cc:
```diff
@@ -0,0 +1,165 @@
+
+private:
+	bool restored = false;
+};
```
Why mocking? Why not use HAFloatingIpManager directly?

ralcolea replied:
Mocking is more appropriate here because `HAFloatingIpManager::restoreFloatingIP()` relies on the `saunafs-uraft-helper` bash script. Therefore, mocking allows us to focus on testing the intended functionality without external dependencies.

uristdwarf commented on src/uraft/main.cc:
```diff
@@ -48,7 +48,10 @@ void parseOptions(int argc, char **argv, uRaftController::Options &opt, bool &ma
+	("URAFT_STATUS_PORT", po::value<int>()->default_value(9428), "node status port")
+	("URAFT_FLOATING_IP", po::value<std::string>(), "floating IP address")
+	("URAFT_FLOATING_IFACE", po::value<std::string>(), "floating IP interface")
+	("URAFT_FLOATING_IP_CHECK_PERIOD", po::value<int>()->default_value(500), "floating IP check status period");
```
One way of allowing to disable this floating IP check is by setting this to 0 and not using the floating IP manager.

uristdwarf replied:
Regardless, the types should be uint, not int

ralcolea replied:
I added the ability of disabling the floating IP manager by setting `URAFT_FLOATING_IP_CHECK_PERIOD=0` in the `saunafs-uraft.cfg`, as you suggested. I also updated the types to use uint.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+		syslog(LOG_WARNING, "Command %s was not successful", command.c_str());
+	}
+
+	return (result == 0);
```
According to `man 3 system`:

```
       •  If  command is NULL, then a nonzero value if a shell is available, or
          0 if no shell is available.

       •  If a child process could not be created, or its status could  not  be
          retrieved,  the  return  value is -1 and errno is set to indicate the
          error.

       •  If a shell could not be executed in the child process, then  the  re‐
          turn  value  is  as  though  the  child  shell  terminated by calling
          _exit(2) with the status 127.

       •  If all system calls succeed, then the return value is the termination
          status of the child shell used to execute command.  (The  termination
          status  of  a  shell is the termination status of the last command it
          executes.)
```

You should check for these cases

ralcolea replied:
I have updated the function to execute the command. Please review the changes and let me know your thoughts.

uristdwarf commented on src/uraft/floating-ip-manager.cc:
```diff
@@ -0,0 +1,228 @@
+
+bool HAFloatingIPManager::restoreFloatingIP() {
+	std::string command = "saunafs-uraft-helper assign-ip";
+	int result = system(command.c_str());
```
This system could get stuck, since there's no timeout

ralcolea replied:
I also considered this requirement in the new version.

uristdwarf reviewed (changes requested):
First off, amazing job you have done with this. This is really great and it's nice to see some tests being added.

However, I have a few requests:

1. In all previous iterations of this feature, we've been forcing users to use this. However in case anything goes wrong, we should allow them to disable this. That means we don't have to critically patch this feature in case anything goes wrong. See one of my comments regarding an idea on how to do this.
2. You are using a lot of standard C library functions. Please double check the manuals for these (with `man 3 <FUNCTION_NAME>`) and make sure you are checking and printing the actual error messages. Make sure you are also freeing any pointers they return if they require it.

rolysr commented on src/uraft/floating_ip_manager_unittest.cc:
```diff
@@ -0,0 +1,165 @@
+		std::exit(EXIT_FAILURE);
+	}
+
+	// Wait for a short period to allow the listener to detect the IP loss
+	std::this_thread::sleep_for(std::chrono::seconds(1));
+
+	// Verify that handleIPLoss was called
+	EXPECT_TRUE(mockManager->isFloatingIPAlive());
```
Consider checking `EXPECT_FALSE(mockManager->isFloatingIPAlive());` before the short period wait to have covered the case when `isFloatingIPAlive` returns false as expected.

ralcolea replied:
Nice catch!!! Done.

rolysr commented on src/uraft/floating_ip_manager_unittest.cc:
```diff
@@ -0,0 +1,165 @@
+		std::exit(EXIT_FAILURE);
+	}
+
+	// Wait for a short period to allow the listener to detect the IP loss
+	std::this_thread::sleep_for(std::chrono::seconds(1));
+
+	// Verify that handleIPLoss was only called for mockManager105
+	EXPECT_TRUE(mockManager105->wasFloatingIPRestored());
+	EXPECT_FALSE(mockManager->wasFloatingIPRestored());
```
Similar consideration from my previous comment, but this time with the  `wasFloatingIPRestored()` returned value

ralcolea replied:
Nice catch!!! Done.

rolysr reviewed (approved):
Well done @ralcolea. LGTM. Just a few comments added.

ralcolea committed:
chore: Apply suggestions from @lgsilva3087 and @rolysr

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
chore: Apply suggestions from @uristdwarf

Signed-off-by: Crash <crash@leil.io>

ralcolea committed:
chore: Reuse `runCommand()` to restore floating IP

The existing functions `runCommand()` and `runSlowCommand()`, correctly
execute `saunafs-uraft-helper` script. Instead of duplicating logic,
this commit refactors `restoreFloatingIp()` to leverage `runCommand()`,
ensuring consistency and reliability.

Key changes:
- Updated documentation to clarify system behavior when the network
  interface used by floating IP is lost.

- Refactored `restoreFloatingIp()` to use `runCommand()` to restore
  the floating IP.

- Added `setCallback()` function to `IHAFloatingIPManager` interface
  to allow setting a function for floating IP restoration.

Signed-off-by: Crash <crash@leil.io>

uristdwarf commented on doc/saunafs-uraft.cfg.5.adoc:
```diff
@@ -46,6 +46,9 @@ server (default: 9421).
 
 *URAFT_FLOATING_IFACE*:: Network interface where floating IP should be managed
 (example: eth0).
+Important: If the network interface used for the floating IP becomes unavailable,
+the failover mechanism will not restore it. As a result, the floating IP cannot
+be recovered, causing SaunaFS services that depend on it to become unavailable.
```
Should mention the failover mechanism is dependent on this option: URAFT_FLOATING_IP_CHECK_PERIOD

ralcolea replied:
Done.
